File: backend/app/schemas/portfolio_risk.py
"""Pydantic v2 schemas for Portfolio Risk Overview."""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict


class BuildingRiskPointRead(BaseModel):
    """Minimal building data for map rendering."""

    building_id: str
    address: str
    city: str
    canton: str
    latitude: float | None = None
    longitude: float | None = None
    score: int
    grade: str
    risk_level: str
    open_actions_count: int = 0
    critical_actions_count: int = 0

    model_config = ConfigDict(from_attributes=True)


class RiskDistributionRead(BaseModel):
    """Count of buildings per evidence grade."""

    grade_a: int = 0
    grade_b: int = 0
    grade_c: int = 0
    grade_d: int = 0
    grade_f: int = 0

    model_config = ConfigDict(from_attributes=True)


class PortfolioRiskOverviewRead(BaseModel):
    """Full portfolio risk overview with summary stats."""

    total_buildings: int
    avg_evidence_score: float
    buildings_at_risk: int
    buildings_ok: int
    worst_building_id: str | None = None
    distribution: RiskDistributionRead
    buildings: list[BuildingRiskPointRead]

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/api/portfolioRisk.ts
import { apiClient } from '@/api/client';

export interface RiskDistribution {
  grade_a: number;
  grade_b: number;
  grade_c: number;
  grade_d: number;
  grade_f: number;
}

export interface BuildingRiskPoint {
  building_id: string;
  address: string;
  city: string;
  canton: string;
  latitude: number | null;
  longitude: number | null;
  score: number;
  grade: string;
  risk_level: string;
  open_actions_count: number;
  critical_actions_count: number;
}

export interface PortfolioRiskOverview {
  total_buildings: number;
  avg_evidence_score: number;
  buildings_at_risk: number;
  buildings_ok: number;
  worst_building_id: string | null;
  distribution: RiskDistribution;
  buildings: BuildingRiskPoint[];
}

export const portfolioRiskApi = {
  getOverview: async (): Promise<PortfolioRiskOverview> => {
    const response = await apiClient.get<PortfolioRiskOverview>('/portfolio/risk-overview');
    return response.data;
  },

  getHeatmap: async (): Promise<BuildingRiskPoint[]> => {
    const response = await apiClient.get<BuildingRiskPoint[]>('/portfolio/risk-heatmap');
    return response.data;
  },
};
